velox/type/Type.cpp

```diff
@@ -611,33 +615,48 @@ bool Type::containsUnknown() const {
 
 namespace {
 
-using CustomTypeFactory =
-    std::function<TypePtr(std::vector<TypePtr> childTypes)>;
-
-std::unordered_map<std::string, CustomTypeFactory>& typeFactories() {
-  static std::unordered_map<std::string, CustomTypeFactory> factories;
+std::unordered_map<std::string, std::unique_ptr<const CustomTypeFactories>>&
+typeFactories() {
+  static std::
+      unordered_map<std::string, std::unique_ptr<const CustomTypeFactories>>
+          factories;
   return factories;
 }
 } // namespace
 
-void registerType(const std::string& name, CustomTypeFactory factory) {
+void registerType(
+    const std::string& name,
+    std::unique_ptr<const CustomTypeFactories> factories) {
   auto uppercaseName = boost::algorithm::to_upper_copy(name);
-  typeFactories().emplace(uppercaseName, factory);
+  typeFactories().emplace(uppercaseName, std::move(factories));
 }
 
 bool typeExists(const std::string& name) {
   auto uppercaseName = boost::algorithm::to_upper_copy(name);
   return typeFactories().count(uppercaseName) > 0;
 }
 
-TypePtr getType(const std::string& name, std::vector<TypePtr> childTypes) {
+auto getTypeEntry(const std::string& name) {
   auto uppercaseName = boost::algorithm::to_upper_copy(name);
-  auto it = typeFactories().find(uppercaseName);
+  return typeFactories().find(uppercaseName);
+}
+
+TypePtr getType(const std::string& name, std::vector<TypePtr> childTypes) {
+  auto it = getTypeEntry(name);
+  if (it == typeFactories().end()) {
```

Comment:
This check repeats in 2 places. It would be simpler if getTypeEntry returns a pointer to CustomTypeFactories or nullptr if not found.

Reply:
Revised the function. The callers would still need to check whether the return value is nullptr though.